### PR TITLE
Adds a notification counter to application context

### LINKED ISSUE #2951: notifications button doesn't unfll after you've read all your notifications
When you read your unread notifications the notifications icon stays filled which was confusing. If the page reloads or you navigate to zooniverse home and back then the icon unfills. If you stay within the project level since the page never reloads the icon stays filled which is confusing. Can you make the icon unfill when you hit the 'read' button.  Maybe for @parrish 

I tried this on Firefox on mac. 

I read my notifications and hit the button:

screen shot 2016-09-02 at 6 51 01 am

then I go to classify

screen shot 2016-09-02 at 6 50 52 am

icon is still filled so it looks like I have notifications when In don't 

 Also a comment on the new suggested form for filling out a github issue for this repo. Maybe for @srallen ?  It's geared towards bug fixes, but I thought this was also a place for feature requests. I'm not sure how I would fill out the suggested form to ask for new features not available because there is no expected behavior. Where do feature requests go?

### PR DESCRIPTION
**Pull request by parrish**
Fixes #2951.

This moves notification counting into a separate class and adds that to the app context.

# Review Checklist

- [x] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [x] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [x] Did you deploy a staging branch? https://add-notifications-context.pfe-preview.zooniverse.org

## Optional

- [ ] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?

### CODE REVIEW

**Comment by parrish** [comment]
Yeah, receiving multiple types of notifications (e.g. "somebody replied to a thread you follow" and "somebody mentioned you") is a known issue.  Never did come up with a good resolution to that.

File: app/partials/app.cjsx [hunk 1]
```diff
@@ -12,35 +13,53 @@ PanoptesApp = React.createClass
...
+
     auth.listen 'change', @handleAuthChange
     generateSessionID()
     @handleAuthChange()
 
   componentWillUnmount: ->
     auth.stopListening 'change', @handleAuthChange
 
+  componentWillReceiveProps: (nextProps) ->
+    @updateNotificationsCount params: nextProps.params
```
Comment (parrish) [inline app/partials/app.cjsx]: Insofar as I know, props only changes when the route changes in this component.  Unless I'm wrong, then that's what the goal is.

File: app/partials/app.cjsx [hunk 2]
```diff
@@ -12,35 +13,53 @@ PanoptesApp = React.createClass
...
     initialLoadComplete: @state.initialLoadComplete
     user: @state.user
     geordi: @geordiLogger
+    notificationsCounter: @state.notificationsCounter
+    unreadNotificationsCount: @state.unreadNotificationsCount
 
   getInitialState: ->
     initialLoadComplete: false
     user: null
+    notificationsCounter: new NotificationsCounter()
```
Comment (parrish) [inline app/partials/app.cjsx]: Happy to move it to props.

File: app/lib/notifications-counter.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    this.setSection().then((section) => {
+      this.countUnread().then(() => {
+        this.loading = false;
+      });
+    });
+  }
+
+  setSection() {
+    if(this.owner && this.name) {
+      let slug = `${this.owner}/${this.name}`;
```
Comment (parrish) [inline app/lib/notifications-counter.js]: Yep.  Changing

**Comment by parrish** [comment]
Updated.  Thanks!

File: app/partials/app.cjsx (hunk 1 above)
Comment (srallen) [inline app/partials/app.cjsx]: Right, that's the only props, but I guess I'm just thinking about if in the future there are any other props, we might want an explicit check, but that doesn't need to happen now.

File: app/partials/app.cjsx (hunk 2 above)
Comment (srallen) [inline app/partials/app.cjsx]: I think I'd prefer it as props unless anyone has a good reason it shouldn't be.

Comment (parrish) [inline app/partials/app.cjsx]: Yep, the counter changed to a prop in a70a3aba.  The unread count is still a state, but that shouldn't need to change.

**Pull request merged**